lucene/core/src/java/org/apache/lucene/util/IntArrayDocIdSet.java

```diff
@@ -67,16 +67,21 @@ public int docID() {
 
     @Override
     public int nextDoc() throws IOException {
-      return doc = docs[++i];
+      return doc = docs[i++];
     }
 
     @Override
     public int advance(int target) throws IOException {
-      i = Arrays.binarySearch(docs, i + 1, length, target);
+      int bound = 1;
+      //given that we use this for small arrays only, this is very unlikely to overflow
+      while(i + bound < length && docs[i + bound] < target) {
+        bound *= 2;
+      }
+      i = Arrays.binarySearch(docs, i + bound / 2, Math.min(i + bound, length), target);
```

Comment:
I think it should be `binarySearch(docs, i + bound / 2, Math.min(i + bound + 1, length), target);`? (maybe we need a test for this special case that the result is found at `i + bound` exactly?)

Reply:
correct, the upper bound is exclusive in Arrays.binarySearch . Funnily enough, this does not seem to cause any problem, as the returned value is `- insertion point - 1` with which we do `i = -1 - i`. If the docId was not in the array, we'd return the docId at the hypothetical position where the docId would be if it was in the array. In reality it is there, and what we return is the proper value still :) I think the test coverage is already good enough, I have corrected this, but I am not sure how to test it given that it did not end up causing any bug.